main/src/com/google/refine/commands/row/GetRowsCommand.java

```diff
@@ -63,54 +62,111 @@ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 import com.google.refine.sorting.SortingRowVisitor;
-import com.google.refine.util.ParsingUtilities;
 import com.google.refine.util.Pool;
 
+/**
+ * Retrieves rows from a project (or importing job).
+ *
+ * Those rows can be requested as either: - the batch of rows starting at a given index (included), up to a certain size
+ * - the batch of rows ending at a given index (excluded), again up to a given size. Filters (defined by facets) and the
+ * row/record mode toggle can also be provided.
+ */
 public class GetRowsCommand extends Command {
 
     protected static class WrappedRow {
 
         @JsonUnwrapped
         protected final Row row;
+
+        /**
+         * The logical row index (meaning the one from the original, unsorted grid)
+         */
         @JsonProperty("i")
         protected final int rowIndex;
         @JsonProperty("j")
         @JsonInclude(Include.NON_NULL)
         protected final Integer recordIndex;
 
-        protected WrappedRow(Row rowOrRecord, int rowIndex, Integer recordIndex) {
+        /**
+         * The row index used for pagination (which can be different from the logical one if a temporary sort is
+         * applied)
+         */
+        @JsonProperty("k")
+        protected final int paginationIndex;
+
+        protected WrappedRow(Row rowOrRecord, int rowIndex, Integer recordIndex, int paginationIndex) {
             this.row = rowOrRecord;
             this.rowIndex = rowIndex;
             this.recordIndex = recordIndex;
+            this.paginationIndex = paginationIndex;
         }
     }
 
     protected static class JsonResult {
 
+        /**
+         * Mode of the engine (row or record based)
+         */
         @JsonProperty("mode")
         protected final Mode mode;
+        /**
+         * Rows in the view
+         */
         @JsonProperty("rows")
         protected final List<WrappedRow> rows;
+        /**
+         * Number of rows selected by the current filter
+         */
         @JsonProperty("filtered")
         protected final int filtered;
+        /**
+         * Total number of rows/records in the unfiltered grid
+         */
         @JsonProperty("total")
         protected final int totalCount;
+        /**
+         * Total number of rows in the unfiltered grid (needed to provide a link to the last page)
+         */
+        @JsonProperty("totalRows")
+        protected final int totalRows;
+
         @JsonProperty("start")
-        protected final int start;
+        @JsonInclude(Include.NON_NULL)
+        protected final Integer start;
+        @JsonProperty("end")
+        @JsonInclude(Include.NON_NULL)
+        protected final Integer end;
         @JsonProperty("limit")
         protected final int limit;
         @JsonProperty("pool")
         protected final Pool pool;
 
+        /**
+         * The value to use as 'end' when fetching the page before this one. Can be null if there is no such page.
+         */
+        @JsonProperty("previousPageId")
```

Comment:
Calling this a page ID seems a little unusual. Is the name left over from a previous iteration?

Reply:
Makes sense, I have renamed it to `previousPageEnd`